lintDirtyModulesOnly:true does NOT skip lint on start (webpack --watch or webpack serve)

- Operating System: Linux
- Node Version: 10.23.0
- NPM Version: 6.14.8
- webpack Version: 4.44.2
- eslint-webpack-plugin Version: 2.4.1

### Expected Behavior

With `lintDirtyModulesOnly: true` and running `webpack --watch` or `webpack serve`, I expect lint to be skipped at the start. And only happen after I edit a file and webpack re-compiles.

### Actual Behavior

Lint warnings and errors are reported at start-up.

```
$ npm run webpack:serve

> eslint-webpack-plugin-test@1.0.0 webpack:serve /home/atte/eslint-webpack-plugin-test
> cross-env NODE_ENV=development webpack serve

ℹ ｢wds｣: Project is running at http://localhost:8080/
ℹ ｢wds｣: webpack output is served from /
ℹ ｢wds｣: Content not from webpack is served from /home/atte/eslint-webpack-plugin-test
⚠ ｢wdm｣: Hash: a74bdf32336d9eca0380
Version: webpack 4.44.2
Time: 835ms
Built at: 01/05/2021 5:11:53 PM
  Asset     Size  Chunks             Chunk Names
main.js  369 KiB    main  [emitted]  main
Entrypoint main = main.js
[0] multi (webpack)-dev-server/client?http://localhost:8080 ./src 40 bytes {main} [built]
[./node_modules/ansi-html/index.js] 4.16 KiB {main} [built]
[./node_modules/html-entities/lib/index.js] 449 bytes {main} [built]
[./node_modules/loglevel/lib/loglevel.js] 8.65 KiB {main} [built]
[./node_modules/url/url.js] 22.8 KiB {main} [built]
[./node_modules/webpack-dev-server/client/index.js?http://localhost:8080] (webpack)-dev-server/client?http://localhost:8080 4.29 KiB {main} [built]
[./node_modules/webpack-dev-server/client/overlay.js] (webpack)-dev-server/client/overlay.js 3.51 KiB {main} [built]
[./node_modules/webpack-dev-server/client/socket.js] (webpack)-dev-server/client/socket.js 1.53 KiB {main} [built]
[./node_modules/webpack-dev-server/client/utils/createSocketUrl.js] (webpack)-dev-server/client/utils/createSocketUrl.js 2.91 KiB {main} [built]
[./node_modules/webpack-dev-server/client/utils/log.js] (webpack)-dev-server/client/utils/log.js 964 bytes {main} [built]
[./node_modules/webpack-dev-server/client/utils/reloadApp.js] (webpack)-dev-server/client/utils/reloadApp.js 1.59 KiB {main} [built]
[./node_modules/webpack-dev-server/client/utils/sendMessage.js] (webpack)-dev-server/client/utils/sendMessage.js 402 bytes {main} [built]
[./node_modules/webpack-dev-server/node_modules/strip-ansi/index.js] (webpack)-dev-server/node_modules/strip-ansi/index.js 161 bytes {main} [built]
[./node_modules/webpack/hot sync ^\.\/log$] (webpack)/hot sync nonrecursive ^\.\/log$ 170 bytes {main} [built]
[./src/index.js] 30 bytes {main} [built]
    + 19 hidden modules

WARNING in 
/home/atte/eslint-webpack-plugin-test/src/index.js
  1:1  warning  Expected indentation of 0 spaces but found 2  indent
  1:3  warning  Unexpected console statement                  no-console

✖ 2 problems (0 errors, 2 warnings)
  0 errors and 1 warning potentially fixable with the `--fix` option.

ℹ ｢wdm｣: Compiled with warnings.
```

### How Do We Reproduce?

Example repo: https://github.com/attekemppila/eslint-webpack-plugin-test

```sh
npm run lint
# eslint warnings in src/index.js

npm run webpack:watch
# eslint warnings in src/index.js - reported at start-up, but should not be

npm run webpack:serve
# eslint warnings in src/index.js - reported at start-up, but should not be
```

--- a/src/index.js
+++ b/src/index.js
@@ -34,7 +34,16 @@ export class ESLintWebpackPlugin {
     // From my testing of compiler.watch() ... compiler.watching is always
     // undefined (webpack 4 doesn't define it either) I'm leaving it out
     // for now.
-    compiler.hooks.watchRun.tapPromise(ESLINT_PLUGIN, this.run);
+    let isFirstRun = this.options.lintDirtyModulesOnly;
+    compiler.hooks.watchRun.tapPromise(ESLINT_PLUGIN, (c) => {
+      if (isFirstRun) {
+        isFirstRun = false;
+
+        return Promise.resolve();
+      }
+
+      return this.run(c);
+    });
   }
 
   /**
@@ -81,7 +90,7 @@ export class ESLintWebpackPlugin {
       // @ts-ignore
       const processModule = (module) => {
         if (module.resource) {
-          const file = module.resource.split('?')[0];
+          const [file] = module.resource.split('?');
 
           if (
             file &&

--- a/test/lint-dirty-modules-only.test.js
+++ b/test/lint-dirty-modules-only.test.js
@@ -1,14 +1,49 @@
+import { join } from 'path';
+import { writeFileSync } from 'fs';
+
+import { removeSync } from 'fs-extra';
+
 import pack from './utils/pack';
 
-describe('lint dirty modules only', () => {
-  it('skips linting on initial run', (done) => {
-    const compiler = pack('error', { lintDirtyModulesOnly: true });
+const target = join(__dirname, 'fixtures', 'lint-dirty-modules-only-entry.js');
 
-    compiler.run((err, stats) => {
+describe('lint dirty modules only', () => {
+  let watch;
+  afterEach(() => {
+    if (watch) {
+      watch.close();
+    }
+    removeSync(target);
+  });
+
+  it('skips linting on initial run', (done) => {
+    writeFileSync(target, 'const foo = false\n');
+
+    let next = firstPass;
+    const compiler = pack('lint-dirty-modules-only', {
+      lintDirtyModulesOnly: true,
+    });
+    watch = compiler.watch({}, (err, stats) => next(err, stats));
+
+    function firstPass(err, stats) {
       expect(err).toBeNull();
       expect(stats.hasWarnings()).toBe(false);
       expect(stats.hasErrors()).toBe(false);
+
+      next = secondPass;
+
+      writeFileSync(target, 'const bar = false;\n');
+    }
+
+    function secondPass(err, stats) {
+      expect(err).toBeNull();
+      expect(stats.hasWarnings()).toBe(false);
+      expect(stats.hasErrors()).toBe(true);
+      const { errors } = stats.compilation;
+      expect(errors.length).toBe(1);
+      const [{ message }] = errors;
+      expect(message).toEqual(expect.stringMatching('no-unused-vars'));
       done();
-    });
+    }
   });
 });

